pinot-core/src/main/java/org/apache/pinot/core/data/manager/realtime/RealtimeTableDataManager.java

```diff
@@ -489,14 +489,16 @@ static void setDefaultTimeValueIfInvalid(TableConfig tableConfig, Schema schema,
 
   @Override
   public void addSegment(ImmutableSegment immutableSegment) {
-    if (isUpsertEnabled()) {
-      handleUpsert(immutableSegment);
-      return;
-    }
+    if (immutableSegment instanceof ImmutableSegmentImpl) {
```

Comment:
@Jackie-Jiang If you recall, I mentioned that the integration test was failing because `RealtimeTableDataManager.handleUpsert` doesn't work with empty segments. When the compacted segment is empty, the call to `SegmentUtils.getRealtimeSegmentPartitionId` returns `null` and causes the precondition to fail. 

If this is preventing the original segment from being replaced by the empty compacted segment, is there a way to ensure that the empty compacted segment has `SegmentPartitionMetadata`?

Reply:
Interesting point. If we are replacing a segment that was a consuming segment, the partition id can be derived from the segment name, which is fine. If the segment was uploaded, then we need to keep the partition in the segment metadata the same as the replaced segment. But we cannot skip the empty segment for upsert handling.

On the other hand, I think we might be able to avoid uploading empty segment. When we find one segment does not contain any valid doc, we can directly delete the segment, instead of uploading an empty segment to replace it. This logic can be applied in the segment generator. In the executor, if we find a segment does not contain any valid doc, we can skip processing it (note that the valid doc id read from executor might not be the same as the one read from the generator).

Reply:
Instead of letting the minion figure out a segment is empty and call controller API to delete the segment, a more straight forward way is to make the task generator (running on controller) to directly delete the segment when it finds there is no valid doc associated with the segment. See the comment above

Reply:
@Jackie-Jiang Thanks for the clarification. Deleting the segments from the generator should be easy as it's already using a PinotHelixResourceManager. But what if the executor gets no validDocIds (eg validDocIds become empty after task is generated)? Should the executor throw a `TaskCancellationException` in this case?

Reply:
It appears that throwing a `TaskCancellationException` causes the task to get stuck in the `IN_PROGRESS` state when the task hasn't timed out. Instead I updated the executor to throw a `FatalException` after deleting the index and working directories.

Reply:
We can just skip processing the task and return success in that case because the generator can remove the segment in the next round

Reply:
The reason I updated the executor to throw an exception is because `BaseSingleSegmentConversionExecutor.executeTask` (which calls `UpsertCompactionTaskExecutor.convert`) requires a valid `SegmentConversionResult` with a file that it **unconditionally** tars and uploads.

Reply:
In `BaseSingleSegmentConversionExecutor`, when CRC does not match, we return a result without setting the `File` (ling 89-95). Similarly, we can return a result without the `File` from `convert()`, and check whether `File` is set after finishing the conversion (line 136). If `File` is not set, directly return.

Reply:
Got it.